### PR TITLE
User table backend

### PR DESCRIPTION
**Pull request by sltimmins**

### CODE REVIEW

**Comment by sltimmins** [comment]
Users backend is complete with /users and /login route complete

**Comment by tmcnitt** [comment]
See slack comments

File: backend/routes.js [hunk 1]
```diff
@@ -95,4 +96,78 @@ module.exports = function routes(app, logger) {
...
+    // obtain a connection with server
+    pool.getConnection(function (err, connection) {
+      if(err) { // if there is an error obtaining a connection
+        logger.error('Problem obtaining MySQL connection', err)
+        res.status(400).send('Problem obtaining MySQL connection');
+      }
+      else { // if there is no error obtaining a connection
+        const username = req.body.username
+        const password = req.body.password
+        const is_construction_firm = JSON.parse(req.body.is_construction_firm)
```
Comment (tmcnitt) [inline backend/routes.js:112]: Why do you json parse just this?

File: backend/routes.js [hunk 2]
```diff
@@ -95,4 +96,78 @@ module.exports = function routes(app, logger) {
...
+        const saltRounds = 10;
+        
+        bcrypt.genSalt(saltRounds, function(err, salt) {
+          bcrypt.hash(password, salt, function(err, hash) {
+            const sql = 'INSERT INTO db.users (username, pass, is_construction_firm) VALUES(?, ?, ?)';
+            connection.query(sql, [username, hash, is_construction_firm], function(err, rows, fields) {
+              connection.release();
+              if(err) {
+                logger.error("Error adding new user: \n", err);
+                res.status(400).send("Username is taken.")
```
Comment (tmcnitt) [inline backend/routes.js:122]: Can you change this and the other error message into a json object..
JSON.stringify("{success: false, msg: 'Username is taken'}")

File: backend/routes.js [hunk 3]
```diff
@@ -95,4 +96,78 @@ module.exports = function routes(app, logger) {
...
+        logger.error('Problem obtaining MySQL connection',err)
+        res.status(400).send('Problem obtaining MySQL connection'); 
+      } else {
+        // if there is no issue obtaining a connection, execute query and release connection
+        const username = req.body.username
+        var sql = "SELECT * FROM db.users WHERE username = ?";
+        connection.query(sql, [username], function (err, rows, fields) {
+          if (err) {
+            logger.error("Error while username salt: \n", err);
+            res.status(400).send("Invalid username.")
```
Comment (tmcnitt) [inline backend/routes.js:147]: JSON.stringify("{success: false, msg: 'Invalid username'}")

File: backend/routes.js [hunk 4]
```diff
@@ -95,4 +96,78 @@ module.exports = function routes(app, logger) {
...
+            
+            bcrypt.compare(password, hash, function(err, result) {
+              if(result) {
+                sql = "SELECT username, is_construction_firm FROM db.users WHERE username = ?";
+                connection.query(sql, [username], function(err, rows, fields) {
+                  if(err) {
+                    logger.error("Error retrieving information: \n", err);
+                    res.status(400).send("Error retrieving login information from database.")
+                  } else {
+                    res.status(200).end(JSON.stringify(rows));
```
Comment (tmcnitt) [inline backend/routes.js:160]: I assume you'll only have one row here?
JSON.stringify("{success: true, user: rows[0]}")

File: backend/routes.js (hunk 1 above)
Comment (sltimmins) [reply backend/routes.js:112]: Oops that was from debugging I'll change it back lol

File: backend/routes.js (hunk 2 above)
Comment (sltimmins) [reply backend/routes.js:122]: Yep

File: backend/routes.js [hunk 5]
```diff
@@ -95,4 +96,78 @@ module.exports = function routes(app, logger) {
...
+                  }
+                });
+              } else {
+                logger.error("Error no matching password: \n", err);
+                res.status(400).send("Incorrect username or password");
```
Comment (tmcnitt) [inline backend/routes.js:165]: JSON.stringify("{success: false, msg: 'Incorrect username or password'}")

**Pull request merged**